### PR TITLE
AC-732 : offline update bug fixed

### PR DESCRIPTION
**Pull request by LuGO0**
## Description of what I changed
Provided a WorkManager Implementation for handling deferred updates to server after modification in the details of a patient.

1. added support for WorkManager in app level Build.gradle work_version ="2.3.3"
2. implemented worker for deferred update operations to the server.
## Issue I worked on
JIRA Issue: https://issues.openmrs.org/browse/AC-732

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains ONE single commit
(the number above, next to the 'Commits' tab is 1).

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by LuGO0** [comment]
@f4ww4z please review the changes .

File: openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                        ToastUtil.success("Patient " + patientTobeUpdated.getPerson().getName().getNameString() + " updated");
+                    }
+                });
+            }
+
+            @Override
+            public void onErrorResponse(String errorMessage) {
+                result[0] = false;
+
+                handler.post(new Runnable() {
```
Comment (AryanGanotra07) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: Have you take care of destroying handler on activity/fragment destroyed?

Comment (LuGO0) [reply openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: @AryanGanotra07 can you please explain what could go wrong with this implementation I would love to patch things up

Comment (LuGO0) [reply openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: btw I was thinking of doing away with these toasts( which require access to UI thread ) but they were earlier a part of update operation thats why I kept them with worker's implementation as well.

Comment (AryanGanotra07) [reply openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: > @AryanGanotra07 can you please explain what could go wrong with this implementation I would love to patch things up

handler.removeCallbacks("your runnable");

Comment (LuGO0) [reply openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: I have used the application context here so where should I destroy this runnable @AryanGanotra07

File: openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            }
+            @Override
+            public void onFailure(@NonNull Call<PatientPhoto> call, @NonNull Throwable t) {
+                ToastUtil.notify("Patient photo cannot be synced due to error: " + t.toString() );
+            }
+        });
+    }
+}
+
+
```
Comment (tbagz104) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: Can you remove the extra spaces? :)

File: openmrs-client/src/main/java/org/openmrs/mobile/api/repository/PatientRepository.java [hunk 3]
```diff
@@ -191,7 +209,7 @@ public void registerPatient(final Patient patient, @Nullable final DefaultRespon
     /**
      * Update Patient
      */
```
Comment (tbagz104) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/repository/PatientRepository.java]: Can you make this into a single line comment and remove the spaces before pushing next time?

File: openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java (hunk 2 above)
Comment (LuGO0) [reply openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java]: Yes sure I would do that once changes are approved thanks for pointing out

File: openmrs-client/src/main/java/org/openmrs/mobile/api/repository/PatientRepository.java (hunk 3 above)
Comment (LuGO0) [reply openmrs-client/src/main/java/org/openmrs/mobile/api/repository/PatientRepository.java]: I dint make those changes @tbagz104

File: openmrs-client/build.gradle [hunk 4]
```diff
@@ -124,7 +124,9 @@ dependencies {
     def room_version = "2.1.0-alpha04"
     def appcompat_version = "1.2.0-alpha01"
     def material_version = "1.2.0-alpha03"
+    def versions_work = "2.3.3"
```
Comment (f4ww4z) [inline openmrs-client/build.gradle:127]: change to 
```suggestion
    def worker_manager_version = "2.3.3"
```

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientActivity.java [hunk 5]
```diff
@@ -82,7 +82,7 @@ protected void onCreate(Bundle savedInstanceState) {
         PlacesClient placesClient = Places.createClient(this);
 
         // Create the mPresenter
-        mPresenter = new AddEditPatientPresenter(addEditPatientFragment, countries, patientID, placesClient);
+        mPresenter = new AddEditPatientPresenter(addEditPatientFragment, countries, patientID, placesClient,getApplicationContext());
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientActivity.java:85]:
```suggestion
        mPresenter = new AddEditPatientPresenter(addEditPatientFragment, countries, patientID, placesClient, getApplicationContext());
```

File: openmrs-client/src/main/java/org/openmrs/mobile/api/repository/PatientRepository.java [hunk 6]
```diff
@@ -61,6 +69,16 @@
...
     private RestApi restApi;
+    private WorkManager mWorkManager=null;
+
+    //constructor specifically for update which uses a workManager Implementation
+    public PatientRepository(Context appContext){
+        this.logger = new OpenMRSLogger();
+        this.patientDao = new PatientDAO();
+        this.locationRepository = new LocationRepository();
+        this.restApi = RestServiceBuilder.createService(RestApi.class);
+        this.mWorkManager=WorkManager.getInstance(appContext);
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/repository/PatientRepository.java:80]: format this file

File: openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        logger = new OpenMRSLogger();
+
+        mHandler = new Handler(Looper.getMainLooper()) {
+            @Override
+            public void handleMessage(Message msg) {
+                String responseMessage;
+                switch (msg.what) {
+                    case ON_SUCCESS:
+                        String UpdateSuccessPatientName = (String) msg.obj;
+                        ToastUtil.success("Patient " + UpdateSuccessPatientName + " updated");
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java:49]: Put this text in `strings.xml` with named parameters. See https://developer.android.com/guide/topics/resources/string-resource.html#formatting-strings . Do this for other toast messages.

File: openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        restApi = RestServiceBuilder.createService(RestApi.class);
+        logger = new OpenMRSLogger();
+
+        mHandler = new Handler(Looper.getMainLooper()) {
+            @Override
+            public void handleMessage(Message msg) {
+                String responseMessage;
+                switch (msg.what) {
+                    case ON_SUCCESS:
+                        String UpdateSuccessPatientName = (String) msg.obj;
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/workers/UpdateWorker.java:48]: Please follow java conventions. 
```suggestion
                        String updateSuccessPatientName = (String) msg.obj;
```

**Comment by LuGO0** [comment]
@f4ww4z I have fixed all the formatting errors and made the requested changes please review .

**Pull request merged**